io/include/traccc/io/csv.hpp

```diff
@@ -235,83 +236,111 @@ inline cell_container_types::host read_cells(
     const Acts::GeometryHierarchyMap<digitization_config>* digi_cfg = nullptr,
     unsigned int max_cells = std::numeric_limits<unsigned int>::max()) {
 
-    cell_container_types::host result(&resource);
+    // Create cell counter vector
+    cell_counter_collection_types::host cell_counts(&resource);
+    // Create cell vector
+    cell_collection_types::host allCells(&resource);
```

Comment:
Annoyingly, [the csv file reading object we're currently](https://eic.github.io/doxygen/da/dae/classdfe_1_1io__dsv__impl_1_1NamedTupleDsvReader.html) using does not allow for getting the number of lines in the file, despite the fact that the actual file reader it holds has this, which could've been used here to immediately reserve the number of cells needed. 
Anyways, it would not make a significant difference to the time taken.

Reply:
I was trying to see if I could find a way of counting the lines in a text file in C\+\+ without having to read the content of the whole file in the process, but didn't find any. :confused: (Thinking a bit more, probably it's not possible to do that...)

So let's just accept this for now. One thing that you could try is to reserve some large-ish number of elements in these vectors. (As we discussed yesterday, using [std::vector::reserve](https://en.cppreference.com/w/cpp/container/vector/reserve).) Since remember, [std::vector](https://en.cppreference.com/w/cpp/container/vector) doubles its allocation every time it runs out of memory. So if we start with let's say 50k elements, the number of re-allocations should stay small even for very large input files.